src/gc/transaction_level_gc_manager.cpp

```diff
@@ -339,139 +278,127 @@ void TransactionLevelGCManager::AddToRecycleMap(
+      recycle_stack->RemoveAllWithTileGroup(tile_group_id);
+      immutable = true;
     }
 
     if (num_recycled >= compaction_threshold) {
-      // TODO: compact this tile group
+      // create task to compact this tile group
+      // add to the worker queue
+      auto &pool = threadpool::MonoQueuePool::GetInstance();
+      pool.SubmitTask([tile_group_id] {
+        TileGroupCompactor::CompactTileGroup(tile_group_id);
+      });
     }
   }
 
-  if (recycling) {
-    // this slot should be recycled, add it back to the recycle queue
-    recycle_queue->Enqueue(location);
+  if (!immutable) {
+    // this slot should be recycled, add it back to the recycle stack
+    recycle_stack->Push(location);
   }
-
-  // Check if tile group should be freed
-  if (num_recycled == tuples_per_tile_group && recycling == false) {
+  // if this is the last remaining tuple recycled, free tile group
+  else if (num_recycled == tuples_per_tile_group) {
     // This GC thread should free the TileGroup
     while (tile_group_header->GetGCReaders() > 1) {
       // Spin here until the other GC threads stop operating on this TileGroup
     }
-    table->DropTileGroup(location.block);
-
-    // TODO: clean the recycle queue of this TileGroup's ItemPointers
-    // RemoveInvalidSlotsFromRecycleQueue(recycle_queue, tile_group_id);
-    // For now, we'll rely on GetRecycledTupleSlot to consume and ignore invalid slots
+    table->DropTileGroup(tile_group_id);
   }
+
   tile_group_header->DecrementGCReaders();
 }
 
-// returns a free tuple slot that can now be recycled/reused, if one exists
-// called by data_table.
-ItemPointer TransactionLevelGCManager::GetRecycledTupleSlot(storage::DataTable *table) {
+// looks for a free tuple slot that can now be reused
+// called by data_table, which passes in a pointer to itself
+ItemPointer TransactionLevelGCManager::GetRecycledTupleSlot(
+    storage::DataTable *table) {
+
   if (table == nullptr) {
     return INVALID_ITEMPOINTER;
   }
-  auto table_id = table->GetOid();
-  std::shared_ptr<peloton::LockFreeQueue<ItemPointer>> recycle_queue;
 
-  if (recycle_queues_->Find(table_id, recycle_queue) == false) {
-    // Table does not have a recycle queue, likely a catalog table
+  auto table_id = table->GetOid();
+  auto recycle_stack = GetTableRecycleStack(table_id);
+  if (recycle_stack == nullptr) {
+    // Table does not have a recycle stack, likely a catalog table
     return INVALID_ITEMPOINTER;
   }
 
-  ItemPointer location;
-  // Search for a slot that can be recycled
-  // TODO: We're relying on GetRecycledTupleSlot to clean the recycle queue. Fix this later.
-  while (recycle_queue->Dequeue(location) == true) {
-    auto tile_group_id = location.block;
-    auto tile_group = table->GetTileGroupById(tile_group_id);
+  // Try to get a slot that can be recycled
+  ItemPointer location = recycle_stack->TryPop();
+  if (location.IsNull()) {
+    return INVALID_ITEMPOINTER;
+  }
 
```

Comment:
I think it may be better for this function GetRecycledTupleSlot() to have a while around TryPop() instead of having while loop around GetRecycledTupleSlot, since this function name seems it guarantees to get a tuple slot.